File: 2015/day03/task2/main.py
with open('input', 'r') as f:
	data = f.read()
loc = [(0, 0), (0, 0)]
visited = set([(0, 0)])
i = 0
for dir in data:
	if dir == '^':
		loc[i] = (loc[i][0] + 1, loc[i][1])
	elif dir == 'v':
		loc[i] = (loc[i][0] - 1, loc[i][1])
	elif dir == '>':
		loc[i] = (loc[i][0], loc[i][1] + 1)
	else:
		loc[i] = (loc[i][0], loc[i][1] - 1)
	visited.add(loc[i])
	i = (i + 1) % 2
result = len(visited)
print(result)

File: 2015/day03/task2/main.clj

(require '[clojure.set :as set])

(def readInput (fn [] (slurp "input")))

(def char2dir (fn [c] (case c
                       \^ [0 1]
                       \v [0 -1]
                       \< [-1 0]
                       \> [1 0]
                       [0 0]
)))

(def data (readInput))

(def moves (vec (map char2dir data)))

(def steps (count moves))

(def addLoc (fn [loc d] [(+ (get loc 0) (get d 0)) (+ (get loc 1) (get d 1))]))

(def visited1 (loop [i 0  loc [0 0] locs (set [[0, 0]])]
                   (if (>= i steps) locs
                   (recur (+ i 2) (addLoc loc (moves i)) (conj locs (addLoc loc (get moves i)))))))
                   
(def visited2 (loop [i 1  loc [0 0] locs (set [[0, 0]])]
                   (if (>= i steps) locs
                   (recur (+ i 2) (addLoc loc (moves i)) (conj locs (addLoc loc (get moves i)))))))
    
(def totalVisited (set/union visited1 visited2))
    
(def result (count totalVisited))
(println result)